File: gaming4free.py
import time
import os
import json
import re
import random
import requests

# 智能环境配置
if "DISPLAY" not in os.environ:
    os.environ["DISPLAY"] = ":1"
    
if "XAUTHORITY" not in os.environ:
    if os.path.exists("/home/headless/.Xauthority"):
        os.environ["XAUTHORITY"] = "/home/headless/.Xauthority"

from seleniumbase import SB

# ================= 配置区域 =================
PROXY_URL = os.getenv("PROXY", "")  
TG_TOKEN = os.getenv("TG_TOKEN")  
TG_CHAT_ID = os.getenv("TG_CHAT_ID")  
SERVERS = os.getenv("SERVERS", "").strip()  

SERVER_LIST = []
if SERVERS:
    for item in SERVERS.split("|"):
        try:
            num, region = item.split(",", 1)
            SERVER_LIST.append({"num": num.strip(), "region": region.strip()})
        except:
            print(f"⚠️ SERVERS 配置格式错误: {item}")
# ===========================================

class Game4FreeRenewal:
    def __init__(self):
        self.BASE_DIR = os.path.dirname(os.path.abspath(__file__))
        self.screenshot_dir = os.path.join(self.BASE_DIR, "artifacts")
        if not os.path.exists(self.screenshot_dir):
            os.makedirs(self.screenshot_dir)

    def log(self, msg):
        timestamp = time.strftime('%H:%M:%S')
        print(f"[{timestamp}] [INFO] {msg}", flush=True)

    def human_wait(self, min_s=6, max_s=10):
        time.sleep(random.uniform(min_s, max_s))

    def time_to_seconds(self, t_str):
        """将 HH:MM:SS 格式转换为秒数，用于严格校验续期是否生效"""
        try:
            h, m, s = map(int, t_str.strip().split(':'))
            return h * 3600 + m * 60 + s
        except:
            return 0

    def clear_blocking_ads(self, sb):
        """清理可能遮挡 Turnstile/提交按钮的广告层（不触碰 Cloudflare iframe）"""
        try:
            removed = sb.execute_script("""
                var sel = 'ins, iframe[src*="google"], iframe[src*="doubleclick"], '
                        + 'div[id^="google_ads"], div[class*="ad-"], div[id^="ad_"], '
                        + 'div[class*="overlay"][style*="z-index"], '
                        + '[id*="sp_message"], .fc-consent-root';
                var nodes = document.querySelectorAll(sel);
                var n = 0;
                for (var i = 0; i < nodes.length; i++) {
                    var el = nodes[i];
                    // 绝不动 Cloudflare / Turnstile 相关节点
                    var html = (el.outerHTML || '').toLowerCase();
                    if (html.indexOf('cloudflare') >= 0 || html.indexOf('turnstile') >= 0 || html.indexOf('cf-') >= 0) {
                        continue;
                    }
                    el.remove();
                    n++;
                }
                return n;
            """)
            if removed:
                self.log(f"🧹 已清理 {removed} 个可能遮挡的广告/浮层节点")
        except Exception:
            pass

    def get_turnstile_token(self, sb):
        """读取 Cloudflare Turnstile 凭证；有值即表示验证已通过"""
        try:
            return sb.execute_script("""
                var selectors = [
                    '[name="cf-turnstile-response"]',
                    'textarea[name="cf-turnstile-response"]',
                    'input[name="cf-turnstile-response"]',
                    '[name="g-recaptcha-response"]'
                ];
                for (var i = 0; i < selectors.length; i++) {
                    var el = document.querySelector(selectors[i]);
                    if (el && el.value && el.value.length > 20) {
                        return el.value;
                    }
                }
                // 某些站点把 token 挂在 turnstile widget 的 data 属性上
                var widget = document.querySelector('[data-sitekey], .cf-turnstile, #cf-turnstile');
                if (widget) {
                    var t = widget.getAttribute('data-response') || widget.getAttribute('data-token');
                    if (t && t.length > 20) return t;
                }
                return '';
            """) or ""
        except Exception:
            return ""

    def has_cloudflare_widget(self, sb):
        """探测页面是否存在 Cloudflare Turnstile 组件"""
        try:
            return bool(sb.execute_script("""
                return !!(
                    document.querySelector('iframe[src*="challenges.cloudflare.com"]')
                    || document.querySelector('iframe[src*="turnstile"]')
                    || document.querySelector('iframe[title*="Cloudflare"]')
                    || document.querySelector('[name="cf-turnstile-response"]')
                    || document.querySelector('.cf-turnstile')
                    || document.querySelector('#cf-turnstile')
                );
            """))
        except Exception:
            return False

    def solve_turnstile(self, sb, server_num):
        """
        解 Cloudflare Turnstile。
        成功返回 True；探测到验证框却始终拿不到 token 时返回 False（fail-closed）。
        未探测到验证框则视为免检，返回 True。
        """
        self.log("📡 开始扫描 Cloudflare 验证框...")
        cf_found = False
        for _ in range(8):
            if self.has_cloudflare_widget(sb):
                cf_found = True
                break
            # 可能已经自动通过
            if self.get_turnstile_token(sb):
                self.log("✅ 已存在 Turnstile 凭证，无需手动点击")
                return True
            time.sleep(1)

        if not cf_found:
            # 再等一小会，部分站点延迟注入 widget
            time.sleep(2)
            if self.get_turnstile_token(sb):
                self.log("✅ 延迟检测到 Turnstile 凭证")
                return True
            if not self.has_cloudflare_widget(sb):
                self.log("✅ 扫描未发现验证框，当前 IP 免检")
                return True
            cf_found = True

        self.log("🛡️ 锁定 Cloudflare 验证框，执行物理点击...")
        self.clear_blocking_ads(sb)

        # 尽量把验证框滚到视口中心，避免 GUI 点击打偏
        try:
            sb.execute_script("""
                var iframe = document.querySelector(
                    'iframe[src*="challenges.cloudflare.com"], iframe[src*="turnstile"], iframe[title*="Cloudflare"]'
                );
                var widget = document.querySelector('.cf-turnstile, #cf-turnstile, [data-sitekey]');
                var target = iframe || widget;
                if (target) {
                    target.scrollIntoView({block: 'center', inline: 'center'});
                } else {
                    var submit = document.querySelector('#vm-submit');
                    if (submit) submit.scrollIntoView({block: 'center'});
                }
            """)
            time.sleep(1)
        except Exception:
            pass

        strategies = [
            ("uc_gui_click_captcha", lambda: sb.uc_gui_click_captcha()),
            ("uc_gui_click_captcha(retry)", lambda: sb.uc_gui_click_captcha(retry=True)),
            ("uc_gui_click_captcha(blind)", lambda: sb.uc_gui_click_captcha(blind=True)),
            ("uc_gui_handle_captcha", lambda: sb.uc_gui_handle_captcha()),
        ]

        for attempt in range(1, 5):
            strategy_name, strategy_fn = strategies[(attempt - 1) % len(strategies)]
            try:
                self.log(f"🖱️ 验证尝试 {attempt}/4 使用策略: {strategy_name}")
                strategy_fn()
            except Exception as e:
                self.log(f"⚠️ 策略 {strategy_name} 异常: {e}")

            # Turnstile 出 token 有时需要数秒
            for wait_i in range(6):
                time.sleep(1.5)
                token = self.get_turnstile_token(sb)
                if token:
                    self.log(f"✅ Turnstile 验证成功（token 长度 {len(token)}，策略 {strategy_name}）")
                    return True

            self.clear_blocking_ads(sb)

        # 失败诊断
        token = self.get_turnstile_token(sb)
        try:
            diag = sb.execute_script("""
                return {
                    hasIframe: !!document.querySelector('iframe[src*="challenges.cloudflare.com"], iframe[src*="turnstile"]'),
                    hasResponseField: !!document.querySelector('[name="cf-turnstile-response"]'),
                    responseLen: (document.querySelector('[name="cf-turnstile-response"]') || {}).value
                        ? document.querySelector('[name="cf-turnstile-response"]').value.length : 0,
                    submitDisabled: (function() {
                        var b = document.querySelector('#vm-submit');
                        return b ? (!!b.disabled || b.getAttribute('aria-disabled') === 'true') : null;
                    })(),
                    bodySnippet: (document.body && document.body.innerText || '').slice(0, 300)
                };
            """)
            self.log(f"🩺 验证失败诊断: {json.dumps(diag, ensure_ascii=False) if isinstance(diag, dict) else diag}")
        except Exception:
            pass

        try:
            fail_shot = f"{self.screenshot_dir}/captcha_fail_{server_num}.png"
            sb.save_screenshot(fail_shot)
            self.log(f"📸 已保存验证失败截图: {fail_shot}")
        except Exception:
            pass

        if token:
            self.log(f"✅ 末次检查拿到 token（长度 {len(token)}）")
            return True

        return False

    def move_mouse_human_advanced(self, sb):
        """生成更复杂的随机鼠标移动轨迹"""
        try:
            time.sleep(random.uniform(0.1, 0.4))
            width = sb.execute_script("return window.innerWidth;")
            height = sb.execute_script("return window.innerHeight;")

            regions = [
                (0.1 * width, 0.1 * height, 0.4 * width, 0.4 * height),
                (0.6 * width, 0.6 * height, 0.9 * width, 0.9 * height),
                (width / 2, height / 2, width / 2, height / 2)
            ]
            num_paths = random.randint(2, 3)

            for _ in range(num_paths):
                target_region = random.choice(regions)
                x_dest = random.randint(int(target_region[0]), int(target_region[2]))
                y_dest = random.randint(int(target_region[1]), int(target_region[3]))
                x_offset = random.randint(-5, 5)
                y_offset = random.randint(-5, 5)

                sb.execute_script(f"""
                    var evt = new MouseEvent("mousemove", {{
                        bubbles: true,
                        cancelable: true,
                        clientX: {x_dest + x_offset},
                        clientY: {y_dest + y_offset}
                    }});
                    document.body.dispatchEvent(evt);
                """)
                time.sleep(random.uniform(0.8, 1.5))
        except:
            pass
    
    def get_remaining_time(self, sb):
        remaining_text = "未知"
        try:
            sb.wait_for_element_visible('#sd-timer', timeout=15)
            time.sleep(1)
            remaining_text = sb.get_text('#sd-timer').strip()
        except Exception as e:
            try:
                remaining_text = sb.execute_script("""
                    var el = document.querySelector('#sd-timer');
                    return el ? el.innerText.trim() : null;
                """)
                if not remaining_text:
                    remaining_text = "未知"
            except:
                remaining_text = "未知"
        return remaining_text

    def send_telegram_notify(self, message, photo_path=None):
        if not TG_TOKEN or not TG_CHAT_ID:
            self.log("⚠️ 未配置 TG_TOKEN，跳过推送。")
            return
        try:
            if photo_path and os.path.exists(photo_path):
                url = f"https://api.telegram.org/bot{TG_TOKEN}/sendPhoto"
                with open(photo_path, 'rb') as f:
                    requests.post(url, data={'chat_id': TG_CHAT_ID, 'caption': message}, files={'photo': f})
            else:
                url = f"https://api.telegram.org/bot{TG_TOKEN}/sendMessage"
                requests.post(url, data={'chat_id': TG_CHAT_ID, 'text': message})
            self.log("✅ TG 推送已发送")
        except Exception as e:
            self.log(f"❌ TG 推送失败: {e}")

    def run_single_server(self, server_num, region):
        URL_APP_PANEL = f"https://gaming4free.net/servers/{server_num}"

        self.log("=" * 40)
        self.log(f"🚀 开始续期 [{region}] ({server_num})")
        
        USER_AGENT = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/138.0.0.0 Safari/537.36"

        with SB(
            uc=True,
            test=True,
            headed=True,
            headless=False,
            xvfb=False,
            chromium_arg=f"--no-sandbox,--disable-dev-shm-usage,--disable-gpu,--window-position=0,0,--start-maximized,--disable-blink-features=AutomationControlled,--disable-infobars,--disable-popup-blocking,--user-agent={USER_AGENT}",
            proxy=PROXY_URL if PROXY_URL else None
        ) as sb:
            try:
                self.log("✅ 浏览器已启动！")

                try:
                    sb.open("https://api.ipify.org?format=json")
                    ip_val = json.loads(re.search(r'\{.*\}', sb.get_text("body")).group(0)).get('ip', 'Unknown')
                    parts = ip_val.split('.')
                    self.log(f"✅ 当前出口 IP: {parts[0]}.{parts[1]}.***.{parts[-1]}")
                except:
                    pass

                self.log(f"📂 正在进入续期面板 [{region}] ...")
                sb.uc_open_with_reconnect(URL_APP_PANEL, reconnect_time=5)
                self.human_wait(8, 12)

                if "login" in sb.get_current_url().lower():
                    raise Exception("登录状态失效或权限被拒绝。")

                cookie_btns = ['//button[contains(., "Continue with Recommended Cookies")]', '//button[contains(., "Accept")]', '//button[contains(., "I Agree")]', '//button[contains(., "Consent")]']
                for btn in cookie_btns:
                    if sb.is_element_present(btn):
                        try:
                            sb.click(btn)
                            break
                        except:
                            pass

                timestamp_before = self.get_remaining_time(sb)
                self.log(f"🕒 续期前剩余运行时间: {timestamp_before}")

                sb.execute_script("window.scrollBy(0,800);")
                self.human_wait(2, 4)
                
                try:
                    self.log("🖱️ 正在迹点击 'VOTE + ADD 90 MIN'...")
                    self.move_mouse_human_advanced(sb)
                    sb.wait_for_element_visible("#sd-vote-btn", timeout=10)
                    sb.click('#sd-vote-btn')
                except Exception as e:
                    raise Exception(f"未找到打开模态框的按钮: {e}")

                self.log("⏳ 观看视频广告...")
                time.sleep(35) 
                
                try:
                    sb.execute_script("document.querySelector('#vm-submit').scrollIntoView({block: 'center'});")
                    time.sleep(1)
                except:
                    pass

                # 解 Cloudflare Turnstile：拿不到 token 就直接判失败（fail-closed），
                # 绝不带着空凭证去提交——那正是"时间没增加"的根因。
                if not self.solve_turnstile(sb, server_num):
                    raise Exception("Cloudflare Turnstile 验证失败：始终未能获取有效凭证，已终止提交以避免无效续期。")

                self.human_wait(2, 4)

                try:
                    self.log("🖱️ 正在点击最终提交按钮 'VOTE — ADDS 90 MINUTES'...")
                    # 提交前再清一次可能盖住按钮的贴片广告（不碰 CF 组件）
                    self.clear_blocking_ads(sb)
                    # 确保按钮不仅可见，还要处于可点击的激活状态（防广告遮挡或倒计时锁定）
                    sb.wait_for_element_clickable("#vm-submit", timeout=15)
                    sb.click('#vm-submit')
                    self.human_wait(8, 12)
                except Exception as e:
                    raise Exception("未能点击最终的确认提交按钮，可能是广告仍未加载完成导致按钮未激活。")

                time.sleep(8)
                
                timestamp_after = self.get_remaining_time(sb)
                self.log(f"🕒 续期后剩余运行时间: {timestamp_after}")

                sec_before = self.time_to_seconds(timestamp_before)
                sec_after = self.time_to_seconds(timestamp_after)
                
                if sec_after > 0 and sec_before > 0:
                    if sec_after <= sec_before + 120:  
                        raise Exception("时间并未增加！人机验证失败或提交请求被服务器拦截。")

                final_screenshot = f"{self.screenshot_dir}/final_success_{server_num}.png"
                sb.save_screenshot(final_screenshot)

                msg = f"✅ [{region}] 续期成功\n🖥️ 编号: {server_num}\n🕒 续期前剩余时间: {timestamp_before}\n🎉 续期后剩余时间: {timestamp_after}"
                self.send_telegram_notify(msg, final_screenshot)

            except Exception as e:
                self.log(f"❌ 运行异常: {e}")
                sb.save_screenshot(f"{self.screenshot_dir}/error_{server_num}.png")
                self.send_telegram_notify(f"❌ [{region}] 执行失败: {e}\n🖥️ 编号: {server_num}", f"{self.screenshot_dir}/error_{server_num}.png")

    def run(self):
        if not SERVER_LIST:
            self.log("❌ 未配置 SERVERS")
            return
        for server in SERVER_LIST:
            self.run_single_server(server["num"], server["region"])


if __name__ == "__main__":
    Game4FreeRenewal().run()

File: gaming4free.js
/**
 * Gaming4Free 自动续期（投票面板版）
 *
 * 过 Cloudflare Turnstile 的核心：puppeteer-real-browser 的 turnstile:true 自动求解，
 * 且其反检测指纹足够真，通常让 CF 直接 invisible 放行、不弹 interactive challenge。
 * 这是 SeleniumBase + pyautogui 物理点击方案（在模态框内坐标不准、被 CF 判可疑）
 * 无法解决的死结所在。
 *
 * 已探明的投票面板真实流程（见 chrome-devtools 调试结论）：
 *   点 #sd-vote-btn → RAMP 激励视频广告 manuallyCreateRewardUi() 的 Promise resolve
 *   后才 adWatched=true 且 openVoteModal() 打开含 Turnstile 的模态框 → Turnstile 通过
 *   后 callback 解禁 #vm-submit → 点提交时 getResponse() 拿 token，空 token 静默丢弃
 *   → POST control.gaming4free.net/api/servers/{slug}/vote → 成功时 #vm-msg 加 'ok' class。
 *
 * 环境变量（沿用旧 Python 版约定，无需改 Secrets）：
 *   PROXY       socks5://127.0.0.1:10808
 *   TG_TOKEN    Telegram bot token
 *   TG_CHAT_ID  Telegram chat id
 *   SERVERS     "num1,region1|num2,region2"
 */

const { connect } = require('puppeteer-real-browser');

const PROXY_URL = process.env.PROXY || '';
const TG_TOKEN = process.env.TG_TOKEN;
const TG_CHAT_ID = process.env.TG_CHAT_ID;
const SERVERS = (process.env.SERVERS || '').trim();

const SERVER_LIST = [];
if (SERVERS) {
    for (const item of SERVERS.split('|')) {
        const idx = item.indexOf(',');
        if (idx > 0) {
            SERVER_LIST.push({
                num: item.slice(0, idx).trim(),
                region: item.slice(idx + 1).trim(),
            });
        } else {
            log(`⚠️ SERVERS 配置格式错误: ${item}`);
        }
    }
}

function log(msg) {
    const t = new Date().toTimeString().slice(0, 8);
    console.log(`[${t}] [INFO] ${msg}`);
}

function sleep(ms) {
    return new Promise((r) => setTimeout(r, ms));
}

function humanWait(minS = 2, maxS = 4) {
    return sleep((minS + Math.random() * (maxS - minS)) * 1000);
}

function timeToSeconds(t) {
    if (!t) return 0;
    const m = String(t).trim().match(/(\d{1,2}):(\d{2}):(\d{2})/);
    if (!m) return 0;
    return (+m[1]) * 3600 + (+m[2]) * 60 + (+m[3]);
}

async function sendTelegram(message) {
    if (!TG_TOKEN || !TG_CHAT_ID) {
        log('⚠️ 未配置 TG_TOKEN，跳过推送。');
        return;
    }
    try {
        const res = await fetch(`https://api.telegram.org/bot${TG_TOKEN}/sendMessage`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ chat_id: TG_CHAT_ID, text: message }),
        });
        if (res.ok) log('✅ TG 推送已发送');
        else log(`❌ TG 推送失败: HTTP ${res.status}`);
    } catch (e) {
        log(`❌ TG 推送失败: ${e.message}`);
    }
}

// 读取 #sd-timer 剩余时间文本
async function getRemainingTime(page) {
    try {
        return await page.evaluate(() => {
            const el = document.querySelector('#sd-timer');
            return el ? el.innerText.trim() : '未知';
        });
    } catch (e) {
        return '未知';
    }
}

// 短路 RAMP 激励广告门槛：覆盖 manuallyCreateRewardUi 为立即 resolve
async function bypassRewardAd(page) {
    try {
        await page.evaluate(() => {
            window.ramp = window.ramp || {};
            if (!window.ramp.que || typeof window.ramp.que.push !== 'function') {
                window.ramp.que = { push: (f) => { try { f(); } catch (e) {} } };
            }
            window.ramp.manuallyCreateRewardUi = () => Promise.resolve();
            window.ramp.spaAddAds = window.ramp.spaAddAds || function () {};
        });
        log('🎟️ 已短路激励广告门槛（manuallyCreateRewardUi 立即放行）');
    } catch (e) {
        log(`⚠️ 激励广告短路注入失败: ${e.message}`);
    }
}

// 投票模态框是否打开
async function isVoteModalOpen(page) {
    try {
        return await page.evaluate(() => {
            const ov = document.querySelector('#vote-overlay');
            if (ov && (ov.classList.contains('open') || getComputedStyle(ov).display !== 'none')) return true;
            const m = document.querySelector('.vote-modal');
            return !!(m && getComputedStyle(m).display !== 'none' && m.offsetParent !== null);
        });
    } catch (e) {
        return false;
    }
}

async function waitForVoteModal(page, timeoutS = 25) {
    for (let i = 0; i < timeoutS; i++) {
        if (await isVoteModalOpen(page)) return true;
        await sleep(1000);
    }
    return false;
}

// 读取 Turnstile token（与页面提交逻辑 getResponse() 一致，隐藏字段兜底）
async function getTurnstileToken(page) {
    try {
        return await page.evaluate(() => {
            try {
                if (window.turnstile && typeof window.turnstile.getResponse === 'function') {
                    const r = window.turnstile.getResponse();
                    if (r && r.length > 20) return r;
                }
            } catch (e) {}
            const el = document.querySelector('[name="cf-turnstile-response"]');
            return el && el.value && el.value.length > 20 ? el.value : '';
        });
    } catch (e) {
        return '';
    }
}

// #vm-submit 是否已解禁（Turnstile callback 权威信号）
async function isSubmitEnabled(page) {
    try {
        return await page.evaluate(() => {
            const b = document.querySelector('#vm-submit');
            return b ? (!b.disabled && b.getAttribute('aria-disabled') !== 'true') : false;
        });
    } catch (e) {
        return false;
    }
}

// 等 Turnstile 被 puppeteer-real-browser 自动求解：轮询 submit 解禁 / token 非空
async function waitTurnstileSolved(page, timeoutS = 60) {
    log('📡 等待 puppeteer-real-browser 自动求解 Turnstile...');
    for (let i = 0; i < timeoutS; i++) {
        if (await isSubmitEnabled(page) || (await getTurnstileToken(page))) {
            const token = await getTurnstileToken(page);
            log(`✅ Turnstile 已通过（submit 解禁，token 长度 ${token.length}）`);
            return true;
        }
        if (i === 20) log('⏳ Turnstile 仍在求解中（可能出现 interactive checkbox，自动求解器处理中）...');
        await sleep(1500);
    }
    return false;
}

async function runSingleServer(num, region) {
    const url = `https://gaming4free.net/servers/${num}`;
    log('='.repeat(40));
    log(`🚀 开始续期 [${region}] (${num})`);

    const args = [
        '--no-sandbox',
        '--disable-setuid-sandbox',
        '--disable-dev-shm-usage',
        '--disable-gpu',
        '--window-size=1280,1200',
    ];
    if (PROXY_URL) args.push(`--proxy-server=${PROXY_URL}`);

    let browser, page;
    try {
        ({ browser, page } = await connect({
            headless: false,
            turnstile: true, // 自动求解 Cloudflare Turnstile
            disableXvfb: true, // 外层 workflow 已用 xvfb-run，避免嵌套冲突
            connectOption: { defaultViewport: null, executablePath: '/usr/bin/google-chrome' },
            args,
        }));
    } catch (e) {
        log(`❌ 浏览器启动失败: ${e.message}`);
        await sendTelegram(`❌ [${region}] 浏览器启动失败: ${e.message}\n🖥️ 编号: ${num}`);
        return;
    }

    try {
        log('✅ 浏览器已启动！');
        await page.setViewport({ width: 1280, height: 1200 });

        // 出口 IP
        try {
            await page.goto('https://api.ipify.org?format=json', { waitUntil: 'domcontentloaded', timeout: 30000 });
            const ip = await page.evaluate(() => {
                try { return JSON.parse(document.body.innerText).ip; } catch (e) { return 'Unknown'; }
            });
            const p = String(ip).split('.');
            if (p.length === 4) log(`✅ 当前出口 IP: ${p[0]}.${p[1]}.***.${p[3]}`);
        } catch (e) { /* 忽略 */ }

        log(`📂 正在进入续期面板 [${region}] ...`);
        await page.goto(url, { waitUntil: 'domcontentloaded', timeout: 60000 });
        await humanWait(6, 9);

        if (page.url().toLowerCase().includes('login')) {
            throw new Error('登录状态失效或权限被拒绝。');
        }

        // 关闭各类 Cookie 同意弹窗
        try {
            await page.evaluate(() => {
                const btns = Array.from(document.querySelectorAll('button, span, a'));
                const b = btns.find((el) => {
                    const t = (el.textContent || '').trim().toLowerCase();
                    return t === 'consent' || t === 'accept' || t === 'i agree' || t.includes('recommended cookies');
                });
                if (b) b.click();
            });
        } catch (e) { /* 忽略 */ }

        const timeBefore = await getRemainingTime(page);
        log(`🕒 续期前剩余运行时间: ${timeBefore}`);

        await page.evaluate(() => window.scrollBy(0, 800));
        await humanWait(2, 4);

        // 先短路激励广告门槛，再点 VOTE
        await bypassRewardAd(page);

        log("🖱️ 正在点击 'VOTE + ADD 90 MIN'...");
        try {
            await page.waitForSelector('#sd-vote-btn', { visible: true, timeout: 10000 });
            await page.click('#sd-vote-btn');
        } catch (e) {
            throw new Error(`未找到打开模态框的按钮: ${e.message}`);
        }

        if (!(await waitForVoteModal(page, 25))) {
            // 兜底：ramp 覆盖可能晚于点击，再补一次并尝试直接打开
            await bypassRewardAd(page);
            try { await page.evaluate(() => { if (typeof openVoteModal === 'function') openVoteModal(); }); } catch (e) {}
            if (!(await waitForVoteModal(page, 15))) {
                throw new Error('投票模态框未能打开（激励广告门槛未通过）。');
            }
        }
        log('✅ 投票模态框已打开');

        // 等 puppeteer-real-browser 自动求解 Turnstile（fail-closed：不过就不提交）
        if (!(await waitTurnstileSolved(page, 60))) {
            const diag = await page.evaluate(() => {
                const b = document.querySelector('#vm-submit');
                const el = document.querySelector('[name="cf-turnstile-response"]');
                return {
                    modalOpen: !!document.querySelector('#vote-overlay.open'),
                    hasIframe: !!document.querySelector('iframe[src*="challenges.cloudflare.com"], iframe[src*="turnstile"]'),
                    submitDisabled: b ? !!b.disabled : null,
                    tokenLen: el && el.value ? el.value.length : 0,
                };
            });
            log(`🩺 验证失败诊断: ${JSON.stringify(diag)}`);
            try { await page.screenshot({ path: `artifacts/captcha_fail_${num}.png` }); } catch (e) {}
            throw new Error('Cloudflare Turnstile 验证失败：submit 始终未解禁且无有效 token，已终止提交。');
        }

        await humanWait(2, 4);

        log("🖱️ 正在点击最终提交按钮 'VOTE — ADDS 90 MINUTES'...");
        try {
            await page.waitForFunction(() => {
                const b = document.querySelector('#vm-submit');
                return b && !b.disabled;
            }, { timeout: 15000 });
            await page.click('#vm-submit');
        } catch (e) {
            throw new Error('未能点击最终的确认提交按钮。');
        }
        await humanWait(6, 9);

        // 用页面自身结果提示 #vm-msg 精确判定（成功带 'ok' class，失败带 'err'）
        let voteResult = null;
        for (let i = 0; i < 10; i++) {
            voteResult = await page.evaluate(() => {
                const m = document.querySelector('#vm-msg');
                if (!m) return null;
                const cls = (m.className || '').toLowerCase();
                const txt = (m.textContent || '').trim();
                if (cls.includes('ok')) return { ok: true, text: txt };
                if (cls.includes('err') || cls.includes('error')) return { ok: false, text: txt };
                if (txt && !['submitting…', 'submitting...'].includes(txt.toLowerCase())) return { pending: true, text: txt };
                return null;
            });
            if (voteResult && (voteResult.ok === true || voteResult.ok === false)) break;
            await sleep(1500);
        }

        if (voteResult && voteResult.ok === false) {
            throw new Error(`服务器拒绝了投票请求: ${voteResult.text}`);
        }

        await sleep(6000);

        const timeAfter = await getRemainingTime(page);
        log(`🕒 续期后剩余运行时间: ${timeAfter}`);

        const secBefore = timeToSeconds(timeBefore);
        const secAfter = timeToSeconds(timeAfter);
        const pageSaidOk = !!(voteResult && voteResult.ok === true);
        const timeIncreased = secAfter > 0 && secBefore > 0 && secAfter > secBefore + 120;

        if (!pageSaidOk && !timeIncreased) {
            const detail = voteResult ? voteResult.text : '无页面提示';
            throw new Error(`时间未增加且未收到成功提示（${detail}）。人机验证或广告校验可能未通过。`);
        }
        if (pageSaidOk) log(`✅ 页面确认投票成功: ${voteResult.text}`);

        try { await page.screenshot({ path: `artifacts/final_success_${num}.png` }); } catch (e) {}

        await sendTelegram(
            `✅ [${region}] 续期成功\n🖥️ 编号: ${num}\n🕒 续期前剩余时间: ${timeBefore}\n🎉 续期后剩余时间: ${timeAfter}`
        );
    } catch (e) {
        log(`❌ 运行异常: ${e.message}`);
        try { await page.screenshot({ path: `artifacts/error_${num}.png` }); } catch (err) {}
        await sendTelegram(`❌ [${region}] 执行失败: ${e.message}\n🖥️ 编号: ${num}`);
    } finally {
        try { await browser.close(); } catch (e) {}
    }
}

async function main() {
    if (!SERVER_LIST.length) {
        log('❌ 未配置 SERVERS');
        return;
    }
    // 确保截图目录存在
    try { require('fs').mkdirSync('artifacts', { recursive: true }); } catch (e) {}

    for (const s of SERVER_LIST) {
        await runSingleServer(s.num, s.region);
    }
}

main();
